R/applyScaleSet.R

```diff
@@ -0,0 +1,33 @@
+#' Apply a ScaleSet
+#'
+#' Applies a ScaleSet to a dataframe of events
+#'
+#' @param scale ScaleSet. Accepts a ScaleSet or scaleSet$scales.
+#' @param data Events data from an FcsFile, i.e. `events@expr`
+#' @export
+#' @examples
+#' \dontrun{
+#' getEvents(experimentId, fcsFileId, destination = 'myFile.fcs')
+#' events = flowCore::read.FCS('myFile.fcs')
+#' data = events@exprs
+#' scaleSet = getScaleSets(experimentId)[1,]
+#' applyScaleSet(scaleSet, data)
+#' }
+applyScaleSet = function(scaleset, data) {
+  if ("experimentId" %in% names(scaleset)) {
+    scaleset = scaleset$scales[[1]]
+  }
+  cols = lapply(1:length(colnames(data)), function (i) { 
+    idx = match(colnames(events)[i], scaleset[,"channelName"])
+    if(!is.na(idx)){
+      return(applyScale(scaleset[idx, "scale"], data[, colnames(data)[i]]))
+    }
+    else {
+      sprintf("No scale in scaleset for '%s' channel",colnames(data)[i])
```

Comment:
Let's print this as an actual warning. Untested:

```R
# before the loop
missing = list()

# in the loop body
} else {
  missing = append(missing, colnames(data)[i])
}

# outside of the loop
if (length(missing) > 0) {
  warn(c("Scales do not exist in the scaleset for the following channels:", missing))
}
```

Reply:
`missing` doesn't get appended inside `lapply` because of the same error/bug/oversight I mentioned in https://github.com/primitybio/cellengine-r-toolkit/issues/56...I'll print the error for the first missing channel

Reply:
Hm, not sure I understand...

Reply:
:laughing: apparently the ellipsis after "#56" gets hidden inside the formatting of the link, so it looked like I couldn't structure a logical English sentence. What I mean is:

I tried to implement this as suggested above, where each missing channel would be appended to the `missing` list, but due to the [reference-caching issue](https://win-vector.com/2017/02/26/iteration-and-closures-in-r/) I mentioned in #56 , `missing` is cleared after each loop and the error message returns an empty list. For now, it throws an error for _any_ single missing channel, instead of _all_ missing channels.

Reply:
Anyway, I slightly changed the behavior to `stop` when there isn't a matching channel; I think this is reasonable behavior; although it would be convenient for the user to get a list of all non-matching channels, it doesn't really make sense to run through the entire dataframe of events, running `applyScale` as we go, only to throw an error at the end. (The alternative would be to return scaled events, but _without_ applying a scale to non-matching channels, but then we have _really_ confusing behavior.)

Reply:
Oh, that's a different issue than by-val/by-ref. I think it works if you use `<<-`:

```diff
diff --git a/R/applyScaleSet.R b/R/applyScaleSet.R
index d1915be..7e70643 100644
--- a/R/applyScaleSet.R
+++ b/R/applyScaleSet.R
@@ -19,16 +19,22 @@ applyScaleSet = function(scaleset, data, clamp_q = FALSE) {
   if ("scales" %in% names(scaleset)) {
     scaleset = scaleset$scales[[1]]
   }
-  cols = lapply(1:length(colnames(data)), function (i) {
-    idx = match(colnames(data)[i], scaleset[,"channelName"])
+
+  missing = list()
+
+  cols = lapply(colnames(data), function (cname) {
+    idx = match(cname, scaleset[,"channelName"])
     if(!is.na(idx)){
-      return(applyScale(scaleset[idx, "scale"], data[, colnames(data)[i]], clamp_q=clamp_q))       
-    }
-    else {
-      stop(c("Scales do not exist in the scaleset for the following channel: ", colnames(data)[i]))
+      return(applyScale(scaleset[idx, "scale"], data[, cname], clamp_q=clamp_q))
+    } else {
+      missing <<- append(missing, cname)
     }
   })

+  if (length(missing) > 0) {
+    rlang::warn(c("Scales do not exist in the scaleset for the following channels:", missing))
+  }
+
   df = as.data.frame(do.call(cbind, cols))
   colnames(df) = colnames(data)
   return(df)
```

```diff
diff --git a/tests/testthat/test-applyScaleSet.R b/tests/testthat/test-applyScaleSet.R
index 27808c6..1533b59 100644
--- a/tests/testthat/test-applyScaleSet.R
+++ b/tests/testthat/test-applyScaleSet.R
@@ -60,10 +60,9 @@ test_that("applies clamped arcsinh scale", {
 })

 test_that("prints correct warning message", {
-  data = data.frame("FSC-A" = c(1, 1, 1, 1, 1, 1), check.names = F)
+  data = data.frame("FSC-A" = c(1, 1), "SSC-A" = c(2, 2), check.names = F)
   # SSC-H: wrong channel name
   scaleSet <- jsonlite::fromJSON('[{"scales":[{"channelName":"SSC-H","scale":{"type":"ArcSinhScale","minimum":-200,"maximum":5000,"cofactor":5}}],"_id":"5d2f8b4b21fd0676fb3a6a8c","experimentId":"5d2f8b4b21fd0676fb3a6a70","name":"Scale Set 1","__v":1,"updated":"2020-10-27T18:38:55.554Z"}]')

-  err <- expect_error(applyScaleSet(scaleSet, data))
-  expect_true(grepl("FSC-A", err$message))
+  expect_warning(applyScaleSet(scaleSet, data), "FSC-A.*SSC-A")
 })
```

Reply:
:pray:  Thanks!

The only other thing I'd say is that if we have
```r
  if (length(missing) > 0) {
    rlang::warn(c("Scales do not exist in the scaleset for the following channels:", missing))
  }

  df = as.data.frame(do.call(cbind, cols))
  colnames(df) = colnames(data)
  return(df)
```

instead of 
```r
  if (length(missing) > 0) {
    rlang::warn(c("Scales do not exist in the scaleset for the following channels:", missing))
  } else {
  df = as.data.frame(do.call(cbind, cols))
  colnames(df) = colnames(data)
  return(df)
  }
```

then this returns a complete dataframe, but where the name of a missing channel fills the column for that channel, like:
```r
FSC-A    FSC-W    SSC-A      Ax488-A        PE-A    PE-TR-A PerCP-Cy55-A
1  FSC-A 97247.88 2.560626     1.363318     4.79299   3.151715      2.18298
2  FSC-A 102743.1 2.863556     1.266519    4.650235   3.038532     1.857759
3  FSC-A 111016.8 2.918169     1.041331    4.608332   2.948592     1.843446
4  FSC-A 101362.9  2.89565     1.066198    4.264754   2.593865     2.005251
```
Is this more desirable than `stop`ping the function?

Reply:
Definitely agreed; changing to a `stop`, as I think that makes the most sense